### PR TITLE
[WIP] vSphere Tutorials

### PR DESCRIPTION
**Pull request by cormachogan**
3 new vSphere tutorials which need review before merging.

1. Velero v1.1 deploy with upstream K8s on vSphere (stateless backup test)
2. Velero v1.1 deploy with PKS on vSphere (stateless backup test)
3. Velero v1.1 stateful backup test (with Cassandra)

### CODE REVIEW

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+## Download and extract Velero v1.1
+
+The [Velero v1.1 image can be found here](https://github.com/heptio/velero/releases/tag/v1.1.0.). Download and extract it to the desktop where you wish to manage your Velero backups, then copy or move the `velero` binary to somewhere in your $PATH.
+
+## Deploy and Configure Minio Object Store as a backup destination
+
+Velero sends metadata about the objects being backed up to an S3 Object Store. If you do not have an S3 Object Store available, Velero provides the manifest file to create a Minio S3 Object Store on your Kubernetes cluster. This means that all Velero backsup can be kept on-premises.
+
+* Note: Stateful backups of applications deployed in Kubernetes on vSphere that use the `restic` plugin for backing up Persistent Volumes send the backup data to the same S3 Object Store.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]:
```suggestion
* Note: Stateful backups of applications deployed in Kubernetes on vSphere that use the `restic` plugin for backing up Persistent Volumes would send the backup data to the same S3 Object Store.
```

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+          path: /var/vcap/data/kubelet/pods
+          type: ""
+        name: host-pods
+
+daemonset.extensions/restic edited
+```
+
+## Deploy a sample application to backup
+
+Velero provide a sample `nginx` application for backup testing. This `nginx` deployment assumes the presence of a LoadBalancer for its Service. PKS supports NSX-T integration, so if this is configured, NSX-T will provide this service for you. If you do not have a Load Balancer as part of your Container Network Interface (CNI), there are some easily configuration ones available to get your started. One example is MetalLb, available [here](https://metallb.universe.tf/).
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]:
```suggestion
Velero provides a sample `nginx` application for backup testing. This `nginx` deployment assumes the presence of a LoadBalancer for its Service. PKS supports NSX-T integration, so if this is configured, NSX-T will provide this service for you. If you do not have a Load Balancer as part of your Container Network Interface (CNI), there are some easily configuration ones available to get your started. One example is MetalLb, available [here](https://metallb.universe.tf/).
```

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+Label selector:  <none>
+
+Restore PVs:  auto
+```
+
+## Verify that the restore succeeded
+
+You can see that the restore has now completed. Check to see if the namespace, DaemonSet and service has been restored using the `kubectl` commands shown previously. One item to note is that the `nginx` service may be restored with a new IP address from the LoadBalancer. This is normal.
+
+Now let’s see if we can successfully reach our `nginx` web server on that IP address. Yes I can! Looks like the restore was successful.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]:
```suggestion
Now let’s see if we can successfully reach our `nginx` web server on that IP address. Yes we can! Looks like the restore was successful.
```

File: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md [hunk 4]
```diff
@@ -0,0 +1,253 @@
+---
+title: Velero v1.1 backing up and restoring apps on vSphere
+# image: https://placehold.it/200x200
+excerpt: Velero is an open source tool to safely back up, recover, and migrate Kubernetes clusters and persistent volumes. It works both on premises and in a public cloud. This post demonstrates how it can be used on vSphere.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md]:
```suggestion
excerpt: A How-To guide to run Velero on vSphere.
```

File: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md [hunk 5]
```diff
@@ -0,0 +1,253 @@
+---
+title: Velero v1.1 backing up and restoring apps on vSphere
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md]:
```suggestion
title: Backup and Restore Stateless Apps on vSphere using Velero v1.1
```

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md [hunk 6]
```diff
@@ -0,0 +1,344 @@
+---
+title: Velero v1.1 backing up and restoring apps on Pivotal Container Service (PKS)
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]:
```suggestion
title: Backup and Restore Apps on Pivotal Container Service (PKS) using  Velero v1.1
```

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+categories: ['kubernetes']
+# Tag should match author to drive author pages
+tags: ['Velero']
+---
+Velero version 1.1 provides support to backup Kubernetes applications deployed on top of PKS, the Pivotal Container Service for vSphere. This post will provide detailed information on how to deploy and configure Velero to backup and restore a stateless application (`nginx`) that is running in a Kubernetes cluster deployed by PKS on vSphere. At this time there is no vSphere plugin for snapshotting stateful applications during a Velero backup. In this case, we rely on a third party program called `restic`. However this post does not include an example of how to backup a stateful application. However this post does not include an example of how to backup a stateful application. That is available in another tutorial which can be found [here](./2019-20-09-Velero-v1-1-Stateful-Backup-vSphere.md).
+
+## Overview of steps
+
+* Download and extract Velero v1.1
+* Deploy and Configure Minio Object store
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]:
```suggestion
* Deploy and Configure a Minio Object store
```

File: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md [hunk 8]
```diff
@@ -0,0 +1,253 @@
...
+categories: ['kubernetes']
+# Tag should match author to drive author pages
+tags: ['Velero']
+---
+Velero version 1.1 provides support to backup Kubernetes applications deployed on vSphere. This post will provide detailed information on how to deploy and configure Velero to backup and restore a stateless application (`nginx`) that is running in Kubernetes on vSphere. At this time there is no vSphere plugin for snapshotting stateful applications on vSphere during a Velero backup. In this case, we rely on a third party program called `restic`. However this post does not include an example of how to backup a stateful application. That is available in another tutorial which can be found [here](./2019-20-09-Velero-v1-1-Stateful-Backup-vSphere.md).
+
+## Overview of steps
+
+* Download and extract Velero v1.1
+* Deploy and Configure Minio Object store
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md]:
```suggestion
* Deploy and Configure a Minio Object store
```

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+* Install Velero using the `velero install` command, ensuring that both `restic` support and a Minio `publicUrl` are included
+* Implement steps to support PKS
+* Run a test backup/restore of a stateless application that has been deployed on Kubernetes that has itself been provisioned by PKS, the Pivotal Container Service
+
+## What this post does not show
+
+* A demonstration on how to do backup/restore of a stateful application (i.e. PVs)
+* The assumption is that the Kubernetes nodes in your cluster have internet access. This guide does not show how to use a local repository for Velero images
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]:
```suggestion
* The assumption is that the Kubernetes nodes in your cluster have internet access in order to pull the Velero images. This guide does not show how to add images using a local repository
```

File: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+* Install Velero using the `velero install` command, ensuring that both `restic` support and a Minio `publicUrl` are included
+* Run a test backup/restore of a stateless application that has been deployed on upstream Kubernetes
+
+## What this post does not show
+
+* How to install Velero v1.1 with PKS, Pivotal Container Service, on vSphere
+* A demonstration on how to do backup/restore of a stateful application (i.e. PVs)
+* The assumption is that the Kubernetes nodes in your cluster have internet access. This guide does not show how to use a local repository for Velero images
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md]:
```suggestion
* TThe assumption is that the Kubernetes nodes in your cluster have internet access in order to pull the Velero images. This guide does not show how to add images using a local repository
```

File: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+## Download and extract Velero v1.1
+
+The [Velero v1.1 image can be found here](https://github.com/heptio/velero/releases/tag/v1.1.0.). Download and extract it to the desktop where you wish to manage your Velero backups, then copy or move the `velero` binary to somewhere in your $PATH.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md]:
```suggestion
The [Velero v1.1 binary can be found here](https://github.com/heptio/velero/releases/tag/v1.1.0.). Download and extract it to the desktop where you wish to manage your Velero backups, then copy or move the `velero` binary to somewhere in your $PATH.
```

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+
+## Download and extract Velero v1.1
+
+The [Velero v1.1 image can be found here](https://github.com/heptio/velero/releases/tag/v1.1.0.). Download and extract it to the desktop where you wish to manage your Velero backups, then copy or move the `velero` binary to somewhere in your $PATH.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]:
```suggestion
The [Velero v1.1 binary can be found here](https://github.com/heptio/velero/releases/tag/v1.1.0.). Download and extract it to the desktop where you wish to manage your Velero backups, then copy or move the `velero` binary to somewhere in your $PATH.
```

File: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+
+## Deploy and Configure Minio Object Store as a backup destination
+
+Velero sends metadata about the objects being backed up to an S3 Object Store. If you do not have an S3 Object Store available, Velero provides the manifest file to create a Minio S3 Object Store on your Kubernetes cluster. This means that all Velero backsup can be kept on-premises.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md]: Besides metadata, we also send the actual contents of the k8s resources. Might be useful to extend the explanation.

File: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+Label selector:  <none>
+
+Restore PVs:  auto
+```
+
+## Verify that the restore succeeded
+
+You can see that the restore has now completed. Check to see if the namespace, DaemonSet and service has been restored using the `kubectl` commands shown previously. One item to note is that the `nginx` service may be restored with a new IP address from the LoadBalancer. This is normal.
+
+Now let’s see if we can successfully reach our `nginx` web server on that IP address. Yes I can! Looks like the restore was successful.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md]:
```suggestion
Now let’s see if we can successfully reach our `nginx` web server on that IP address. Yes we can! Looks like the restore was successful.
```

File: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+
+![nginx restored](../img/vsphere-tutorial-icons/nginx-restore-new-ip.png)
+
+Backups and Restore are now working on Kubernetes deployed on vSphere using Velero v1.1.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md]:
```suggestion
Backups and Restores are now working on Kubernetes deployed on vSphere using Velero v1.1.
```

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+
+## Verify that the restore succeeded
+
+You can see that the restore has now completed. Check to see if the namespace, DaemonSet and service has been restored using the `kubectl` commands shown previously. One item to note is that the `nginx` service may be restored with a new IP address from the LoadBalancer. This is normal.
+
+Now let’s see if we can successfully reach our `nginx` web server on that IP address. Yes I can! Looks like the restore was successful.
+
+![nginx restored](../img/vsphere-tutorial-icons/nginx-restore-new-ip.png)
+
+Backups and Restore are now working on Kubernetes deployed by PKS on vSphere using Velero v1.1.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]:
```suggestion
Backups and Restores are now working on Kubernetes deployed by PKS on vSphere using Velero v1.1.
```

File: site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md [hunk 17]
```diff
@@ -0,0 +1,452 @@
...
+
+ emp_id | emp_city | emp_name | emp_phone | emp_sal
+--------+----------+----------+-----------+---------
+    100 |     Cork |   Cormac |       999 | 1000000
+
+(1 rows)
+cqlsh:demodb>
+```
+
+It looks like the restore has been successful. Velero v1.1 has successfully restore the Kubenetes objects for the Cassandra applicaation, as well as restore database and table contents.
```
Comment (carlisia) [inline site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md]:
```suggestion
It looks like the restore has been successful. Velero v1.1 has successfully restored the Kubenetes objects for the Cassandra application, as well as restored the database and table contents.
```

File: site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md [hunk 18]
```diff
@@ -0,0 +1,452 @@
+---
+title: Velero v1.1 backing up and restoring apps on Pivotal Container Service (PKS)
```
[the remark this answers is not in the document]
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md]: LGTM

File: site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md [hunk 19]
```diff
@@ -0,0 +1,452 @@
...
+# image: https://placehold.it/200x200
+excerpt: Velero is an open source tool to safely back up, recover, and migrate Kubernetes clusters and persistent volumes. It works both on premises and in a public cloud. This post demonstrates how it can be used on PKS, the Pivotal Container Service which deploys Kubernetes clusters on vSphere.
+author_name: Cormac Hogan
+# author_avatar: https://placehold.it/64x64
+categories: ['kubernetes']
+# Tag should match author to drive author pages
+tags: ['Velero']
```
[the remark this answers is not in the document]
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md]: Done

File: site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md [hunk 20]
```diff
@@ -0,0 +1,452 @@
...
+---
+Velero version 1.1 provides support to backup Kubernetes applications deployed on upstream Kubernetes deployed natively on vSphere, or deployed on Kubernetes deployed by PKS, the Pivotal Container Service for vSphere. This post will provide detailed information on how to use Velero v1.1 to backup and restore a stateful application (`Cassandra`) that is running in a Kubernetes cluster deployed on vSphere. At this time there is no vSphere plugin for snapshotting stateful applications during a Velero backup. In this case, we rely on a third party program called `restic`.
```
[the remark this answers is not in the document]
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md]: Done

File: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md (hunk 13 above)
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md]: Done

File: site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md [hunk 21]
```diff
@@ -0,0 +1,452 @@
+---
+title: Velero v1.1 backing up and restoring apps on Pivotal Container Service (PKS)
+# image: https://placehold.it/200x200
```
[the remark this answers is not in the document]
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md]: Done

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md (hunk 9 above)
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]: Fixed

File: site/_posts/2019-09-20-Velero-v1-1-on-PKS.md (hunk 7 above)
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-on-PKS.md]: Fixed

File: site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md [hunk 22]
```diff
@@ -0,0 +1,452 @@
...
+excerpt: Velero is an open source tool to safely back up, recover, and migrate Kubernetes clusters and persistent volumes. It works both on premises and in a public cloud. This post demonstrates how it can be used on PKS, the Pivotal Container Service which deploys Kubernetes clusters on vSphere.
```
[the remark this answers is not in the document]
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md]: Done

File: site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md [hunk 23]
```diff
@@ -0,0 +1,452 @@
...
+
+Repeat this action for the other Pods, in this example, Pods cassandra-1 and cassandra-2. This is an indicatation that we need to backup the persistent volume contents associated with each Pod.
+
+## Take a backup
+
+```bash
+$ velero backup create cassandra --include-namespaces cassandra
+Backup request "cassandra" submitted successfully.
+Run `velero backup describe cassandra` or `velero backup logs cassandra` for more details.
+```
```
[the remark this answers is not in the document]
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md]: Done

File: site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md [hunk 24]
```diff
@@ -0,0 +1,452 @@
...
+$ kubectl -n cassandra annotate pod/cassandra-0 backup.velero.io/backup-volumes=cassandra-data
+pod/cassandra-0 annotated
+```
+
+```bash
+$ kubectl -n cassandra describe pod/cassandra-0 | grep Annotations
+Annotations:        backup.velero.io/backup-volumes: cassandra-data
+```
+
+Repeat this action for the other Pods, in this example, Pods cassandra-1 and cassandra-2. This is an indicatation that we need to backup the persistent volume contents associated with each Pod.
```
[the remark this answers is not in the document]
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md]: Fixed

File: site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md [hunk 25]
```diff
@@ -0,0 +1,452 @@
...
+* Verify that the database and table contents have been restored
+
+## What this post does not show
+
+* This tutorial does not show how to deploy Velero v1.1 on vSphere (with or without PKS). These are available in other tutorials.
+* The assumption is that the Kubernetes nodes in your cluster have internet access. This guide does not show how to use a local repository for Velero images.
+
+## Download and Deploy Cassandra
+
+For instructions on how to download and deploy a simple Cassandra StatefulSet, please refer to [this blog post by Cormac Hogan](https://cormachogan.com/2019/06/12/kubernetes-storage-on-vsphere-101-statefulset/). This will show you how to deploy a Cassandra StatefulSet which we can use to do our Stateful aplication backup and restore. The manifests [available here](https://github.com/cormachogan/vsphere-storage-101/tree/master/StatefulSets) use an earlier version of Cassandra (v11) that includes the `cqlsh` tool which we will use now to create a database and populate a table with some sample data.
```
[the remark this answers is not in the document]
Comment (cormachogan) [reply site/_posts/2019-09-20-Velero-v1-1-Stateful-Backup-vSphere.md]: Done

**Pull request closed without merging**